File: leetcode/go/876.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func middleNode(head *ListNode) *ListNode {
	middle := head

	for head != nil && head.Next != nil {
		middle = middle.Next
		head = head.Next.Next
	}

	return middle
}

File: leetcode/C++/876.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
  ListNode *middleNode(ListNode *head) {
    vector<ListNode *> v;

    while (head != NULL) {
      v.push_back(head);
      head = head->next;
    }

    int index = floor(v.size() / 2);

    return v[index];
  }
};
